File: frontend/src/services/automation.ts
import { apiClient } from "../utils/apiClient";
import type {
    PlaybookCreateInput,
    PlaybookUpdateInput,
    PlaybookListData,
    PlaybookData,
    PlaybookStepCreateInput,
    PlaybookStepUpdateInput,
    PlaybookStepData,
    PlaybookStepListData,
    PlaybookVariableCreateInput,
    PlaybookVariableUpdateInput,
    PlaybookVariableData,
    PlaybookVariableListData,
    PlaybookExecuteInput,
    PlaybookExecutionListData,
    PlaybookExecutionData,
    DryRunInput,
    DryRunData,
    RollbackInput,
    RollbackData,
    ExecutionLogListData,
    ExecutionLogData,
    ApprovalWorkflowCreateInput,
    ApprovalWorkflowListData,
    ApprovalWorkflowData,
    ApprovalRequestListData,
    ApprovalRequestData,
    ApprovalActionInput,
    AuditTrailData,
    AuditTrailListData,
    PlaybookScheduleCreateInput,
    PlaybookScheduleListData,
    PlaybookScheduleData,
    EventTriggerCreateInput,
    EventTriggerListData,
    EventTriggerData,
    AutomationSummaryData,
} from "../types/automation";

export type {
    PlaybookCreateInput,
    PlaybookUpdateInput,
    PlaybookListData,
    PlaybookData,
    PlaybookStepCreateInput,
    PlaybookStepUpdateInput,
    PlaybookStepData,
    PlaybookStepListData,
    PlaybookVariableCreateInput,
    PlaybookVariableUpdateInput,
    PlaybookVariableData,
    PlaybookVariableListData,
    PlaybookExecuteInput,
    PlaybookExecutionListData,
    PlaybookExecutionData,
    DryRunInput,
    DryRunData,
    RollbackInput,
    RollbackData,
    ExecutionLogListData,
    ExecutionLogData,
    ApprovalWorkflowCreateInput,
    ApprovalWorkflowListData,
    ApprovalWorkflowData,
    ApprovalRequestListData,
    ApprovalRequestData,
    ApprovalActionInput,
    AuditTrailData,
    AuditTrailListData,
    PlaybookScheduleCreateInput,
    PlaybookScheduleListData,
    PlaybookScheduleData,
    EventTriggerCreateInput,
    EventTriggerListData,
    EventTriggerData,
    AutomationSummaryData,
} from "../types/automation";

const API = "/api/v1/automation";

// ------------------------------------------------------------------ //
// API Client                                                            //
// ------------------------------------------------------------------ //

export const automationApi = {
    // -- Dashboard --
    async getSummary(): Promise<AutomationSummaryData> {
        return apiClient<AutomationSummaryData>(`${API}/dashboard`);
    },

    // -- Playbooks --
    async listPlaybooks(
        search?: string,
        category?: string,
    ): Promise<PlaybookListData> {
        const params = new URLSearchParams();
        if (search) params.set("search", search);
        if (category) params.set("category", category);
        const qs = params.toString();
        return apiClient<PlaybookListData>(
            `${API}/playbooks${qs ? `?${qs}` : ""}`
        );
    },

    async getPlaybook(id: number): Promise<PlaybookData> {
        return apiClient<PlaybookData>(`${API}/playbooks/${id}`);
    },

    async createPlaybook(
        data: PlaybookCreateInput,
    ): Promise<PlaybookData> {
        return apiClient<PlaybookData>(`${API}/playbooks`, {
            method: "POST",
            json: data,
        });
    },

    async updatePlaybook(
        id: number,
        data: PlaybookUpdateInput,
    ): Promise<PlaybookData> {
        return apiClient<PlaybookData>(`${API}/playbooks/${id}`, {
            method: "PUT",
            json: data,
        });
    },

    async deletePlaybook(id: number): Promise<void> {
        return apiClient<void>(`${API}/playbooks/${id}`, {
            method: "DELETE",
        });
    },

    // -- Steps --
    async listSteps(
        playbookId: number,
    ): Promise<PlaybookStepListData> {
        return apiClient<PlaybookStepListData>(
            `${API}/playbooks/${playbookId}/steps`
        );
    },

    async createStep(
        playbookId: number,
        data: PlaybookStepCreateInput,
    ): Promise<PlaybookStepListData> {
        return apiClient<PlaybookStepListData>(
            `${API}/playbooks/${playbookId}/steps`,
            {
                method: "POST",
                json: data,
            }
        );
    },

    async updateStep(
        stepId: number,
        data: PlaybookStepUpdateInput,
    ): Promise<PlaybookStepListData> {
        return apiClient<PlaybookStepListData>(`${API}/steps/${stepId}`, {
            method: "PUT",
            json: data,
        });
    },

    async deleteStep(stepId: number): Promise<void> {
        return apiClient<void>(`${API}/steps/${stepId}`, {
            method: "DELETE",
        });
    },

    // -- Variables --
    async listVariables(
        playbookId: number,
    ): Promise<PlaybookVariableListData> {
        return apiClient<PlaybookVariableListData>(
            `${API}/playbooks/${playbookId}/variables`
        );
    },

    async createVariable(
        playbookId: number,
        data: PlaybookVariableCreateInput,
    ): Promise<PlaybookVariableListData> {
        return apiClient<PlaybookVariableListData>(
            `${API}/playbooks/${playbookId}/variables`,
            {
                method: "POST",
                json: data,
            }
        );
    },

    async updateVariable(
        variableId: number,
        data: PlaybookVariableUpdateInput,
    ): Promise<PlaybookVariableListData> {
        return apiClient<PlaybookVariableListData>(
            `${API}/variables/${variableId}`,
            {
                method: "PUT",
                json: data,
            }
        );
    },

    async deleteVariable(variableId: number): Promise<void> {
        return apiClient<void>(
            `${API}/variables/${variableId}`,
            { method: "DELETE" }
        );
    },

    // -- Executions --
    async listExecutions(
        playbookId?: number,
        statusFilter?: string,
    ): Promise<PlaybookExecutionListData> {
        const params = new URLSearchParams();
        if (playbookId) params.set("playbook_id", String(playbookId));
        if (statusFilter) params.set("status", statusFilter);
        const qs = params.toString();
        return apiClient<PlaybookExecutionListData>(
            `${API}/executions${qs ? `?${qs}` : ""}`
        );
    },

    async getExecution(
        id: number,
    ): Promise<PlaybookExecutionData> {
        return apiClient<PlaybookExecutionData>(`${API}/executions/${id}`);
    },

    async executePlaybook(
        playbookId: number,
        data: PlaybookExecuteInput,
    ): Promise<PlaybookExecutionData> {
        return apiClient<PlaybookExecutionData>(
            `${API}/playbooks/${playbookId}/execute`,
            {
                method: "POST",
                json: data,
            }
        );
    },

    async dryRun(
        playbookId: number,
        data: DryRunInput,
    ): Promise<DryRunData> {
        return apiClient<DryRunData>(
            `${API}/playbooks/${playbookId}/dry-run`,
            {
                method: "POST",
                json: data,
            }
        );
    },

    async rollback(
        executionId: number,
        data: RollbackInput,
    ): Promise<RollbackData> {
        return apiClient<RollbackData>(
            `${API}/executions/${executionId}/rollback`,
            {
                method: "POST",
                json: data,
            }
        );
    },

    // -- Execution Logs --
    async listExecutionLogs(
        executionId: number,
    ): Promise<ExecutionLogListData> {
        return apiClient<ExecutionLogListData>(
            `${API}/executions/${executionId}/logs`
        );
    },

    // -- Approval Workflows --
    async listWorkflows(
        playbookId: number,
    ): Promise<ApprovalWorkflowListData> {
        return apiClient<ApprovalWorkflowListData>(
            `${API}/playbooks/${playbookId}/workflows`
        );
    },

    async createWorkflow(
        playbookId: number,
        data: ApprovalWorkflowCreateInput,
    ): Promise<ApprovalWorkflowListData> {
        return apiClient<ApprovalWorkflowListData>(
            `${API}/playbooks/${playbookId}/workflows`,
            {
                method: "POST",
                json: data,
            }
        );
    },

    async deleteWorkflow(workflowId: number): Promise<void> {
        return apiClient<void>(
            `${API}/workflows/${workflowId}`,
            { method: "DELETE" }
        );
    },

    // -- Approval Requests --
    async listPendingApprovals(): Promise<ApprovalRequestListData> {
        return apiClient<ApprovalRequestListData>(`${API}/approvals/pending`);
    },

    async listExecutionApprovals(
        executionId: number,
    ): Promise<ApprovalRequestListData> {
        return apiClient<ApprovalRequestListData>(
            `${API}/executions/${executionId}/approvals`
        );
    },

    async approveRequest(
        requestId: number,
        data: ApprovalActionInput,
    ): Promise<ApprovalRequestData> {
        return apiClient<ApprovalRequestData>(
            `${API}/approvals/${requestId}/approve`,
            {
                method: "PUT",
                json: data,
            }
        );
    },

    async rejectRequest(
        requestId: number,
        data: ApprovalActionInput,
    ): Promise<ApprovalRequestData> {
        return apiClient<ApprovalRequestData>(
            `${API}/approvals/${requestId}/reject`,
            {
                method: "PUT",
                json: data,
            }
        );
    },

    // -- Audit Trail --
    async listAuditTrail(
        entityType?: string,
        action?: string,
    ): Promise<AuditTrailListData> {
        const params = new URLSearchParams();
        if (entityType) params.set("entity_type", entityType);
        if (action) params.set("action", action);
        const qs = params.toString();
        return apiClient<AuditTrailListData>(
            `${API}/audit${qs ? `?${qs}` : ""}`
        );
    },

    // -- Schedules --
    async listSchedules(
        playbookId?: number,
    ): Promise<PlaybookScheduleListData> {
        const params = new URLSearchParams();
        if (playbookId) params.set("playbook_id", String(playbookId));
        const qs = params.toString();
        return apiClient<PlaybookScheduleListData>(
            `${API}/schedules${qs ? `?${qs}` : ""}`
        );
    },

    async createSchedule(
        playbookId: number,
        data: PlaybookScheduleCreateInput,
    ): Promise<PlaybookScheduleData> {
        return apiClient<PlaybookScheduleData>(
            `${API}/playbooks/${playbookId}/schedules`,
            {
                method: "POST",
                json: data,
            }
        );
    },

    async deleteSchedule(scheduleId: number): Promise<void> {
        return apiClient<void>(
            `${API}/schedules/${scheduleId}`,
            { method: "DELETE" }
        );
    },

    // -- Event Triggers --
    async listTriggers(
        playbookId?: number,
    ): Promise<EventTriggerListData> {
        const params = new URLSearchParams();
        if (playbookId) params.set("playbook_id", String(playbookId));
        const qs = params.toString();
        return apiClient<EventTriggerListData>(
            `${API}/triggers${qs ? `?${qs}` : ""}`
        );
    },

    async createTrigger(
        playbookId: number,
        data: EventTriggerCreateInput,
    ): Promise<EventTriggerData> {
        return apiClient<EventTriggerData>(
            `${API}/playbooks/${playbookId}/triggers`,
            {
                method: "POST",
                json: data,
            }
        );
    },

    async deleteTrigger(triggerId: number): Promise<void> {
        return apiClient<void>(
            `${API}/triggers/${triggerId}`,
            { method: "DELETE" }
        );
    },

    // -- Clone / Export / Import --
    async clonePlaybook(
        playbookId: number,
        name?: string,
    ): Promise<PlaybookData> {
        const params = new URLSearchParams();
        if (name) params.set("name", name);
        const qs = params.toString();
        return apiClient<PlaybookData>(
            `${API}/playbooks/${playbookId}/clone${qs ? `?${qs}` : ""}`,
            { method: "POST" }
        );
    },

    async exportPlaybook(
        playbookId: number,
    ): Promise<{ playbook: unknown; steps: unknown[]; variables: unknown[] }> {
        return apiClient(
            `${API}/playbooks/${playbookId}/export`
        );
    },

    async importPlaybook(
        data: { playbook: unknown; steps: unknown[]; variables: unknown[] },
    ): Promise<PlaybookData> {
        return apiClient<PlaybookData>(
            `${API}/playbooks/import`,
            {
                method: "POST",
                json: data,
            }
        );
    },
};

File: backend/app/routers/automation.py
"""
Mission Control Automation Router

API endpoints for playbooks, steps, executions, approvals,
scheduling, triggers, audit trails, rollback, and dry run.

Sprint 2.8 - Automation & Playbooks.
"""

import logging

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.orm import Session

from app.core.auth_dependency import get_current_user
from app.db import get_db
from app.schemas.approval import (
    ApprovalAction,
    ApprovalRequestListResponse,
    ApprovalRequestResponse,
    ApprovalWorkflowCreate,
    ApprovalWorkflowListResponse,
    ApprovalWorkflowResponse,
    ApprovalWorkflowUpdate,
)
from app.schemas.audit_trail import AuditTrailListResponse
from app.schemas.event_trigger import (
    EventTriggerCreate,
    EventTriggerListResponse,
    EventTriggerResponse,
    EventTriggerUpdate,
)
from app.schemas.execution_log import ExecutionLogListResponse
from app.schemas.playbook import (
    PlaybookCreate,
    PlaybookListResponse,
    PlaybookResponse,
    PlaybookUpdate,
)
from app.schemas.playbook_execution import (
    DryRunRequest,
    DryRunResponse,
    PlaybookExecuteRequest,
    PlaybookExecuteResponse,
    PlaybookExecutionListResponse,
    RollbackRequest,
    RollbackResponse,
)
from app.schemas.playbook_schedule import (
    PlaybookScheduleCreate,
    PlaybookScheduleListResponse,
    PlaybookScheduleResponse,
    PlaybookScheduleUpdate,
)
from app.schemas.playbook_step import (
    PlaybookStepCreate,
    PlaybookStepListResponse,
    PlaybookStepResponse,
    PlaybookStepUpdate,
)
from app.schemas.playbook_variable import (
    PlaybookVariableCreate,
    PlaybookVariableListResponse,
    PlaybookVariableResponse,
    PlaybookVariableUpdate,
)
from app.services.automation_service import (
    AutomationService,
    automation_service,
)

logger = logging.getLogger(__name__)

router = APIRouter(
    prefix="/automation",
    tags=["Automation"],
    dependencies=[Depends(get_current_user)],
)


def get_automation_service() -> AutomationService:
    return automation_service


# ------------------------------------------------------------------ #
# Dashboard Summary                                                    #
# ------------------------------------------------------------------ #


@router.get("/dashboard")
async def get_automation_dashboard(
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> dict:
    return await service.get_automation_summary(db)


# ------------------------------------------------------------------ #
# Playbook Endpoints                                                   #
# ------------------------------------------------------------------ #


@router.get("/playbooks", response_model=PlaybookListResponse)
async def list_playbooks(
    search: str | None = Query(None),
    category: str | None = Query(None),
    enabled: bool | None = Query(None),
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> PlaybookListResponse:
    return await service.get_playbooks(db, search, category, enabled)


@router.get(
    "/playbooks/{playbook_id}",
    response_model=PlaybookResponse,
)
async def get_playbook(
    playbook_id: int,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> PlaybookResponse:
    return await service.get_playbook_by_id(db, playbook_id)


@router.post(
    "/playbooks",
    status_code=status.HTTP_201_CREATED,
    response_model=PlaybookResponse,
)
async def create_playbook(
    payload: PlaybookCreate,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> PlaybookResponse:
    return await service.create_playbook(db, payload)


@router.put(
    "/playbooks/{playbook_id}",
    response_model=PlaybookResponse,
)
async def update_playbook(
    playbook_id: int,
    payload: PlaybookUpdate,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> PlaybookResponse:
    return await service.update_playbook(db, playbook_id, payload)


@router.delete(
    "/playbooks/{playbook_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def delete_playbook(
    playbook_id: int,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> None:
    await service.delete_playbook(db, playbook_id)


# ------------------------------------------------------------------ #
# Playbook Steps                                                       #
# ------------------------------------------------------------------ #


@router.get(
    "/playbooks/{playbook_id}/steps",
    response_model=PlaybookStepListResponse,
)
async def list_playbook_steps(
    playbook_id: int,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> PlaybookStepListResponse:
    return await service.get_playbook_steps(db, playbook_id)


@router.post(
    "/playbooks/{playbook_id}/steps",
    status_code=status.HTTP_201_CREATED,
    response_model=PlaybookStepResponse,
)
async def create_playbook_step(
    playbook_id: int,
    payload: PlaybookStepCreate,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> PlaybookStepResponse:
    return await service.create_playbook_step(
        db, playbook_id, payload
    )


@router.put(
    "/steps/{step_id}",
    response_model=PlaybookStepResponse,
)
async def update_playbook_step(
    step_id: int,
    payload: PlaybookStepUpdate,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> PlaybookStepResponse:
    return await service.update_playbook_step(db, step_id, payload)


@router.delete(
    "/steps/{step_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def delete_playbook_step(
    step_id: int,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> None:
    await service.delete_playbook_step(db, step_id)


# ------------------------------------------------------------------ #
# Playbook Variables                                                   #
# ------------------------------------------------------------------ #


@router.get(
    "/playbooks/{playbook_id}/variables",
    response_model=PlaybookVariableListResponse,
)
async def list_playbook_variables(
    playbook_id: int,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> PlaybookVariableListResponse:
    return await service.get_playbook_variables(db, playbook_id)


@router.post(
    "/playbooks/{playbook_id}/variables",
    status_code=status.HTTP_201_CREATED,
    response_model=PlaybookVariableResponse,
)
async def create_playbook_variable(
    playbook_id: int,
    payload: PlaybookVariableCreate,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> PlaybookVariableResponse:
    return await service.create_playbook_variable(
        db, playbook_id, payload
    )


@router.put(
    "/variables/{variable_id}",
    response_model=PlaybookVariableResponse,
)
async def update_playbook_variable(
    variable_id: int,
    payload: PlaybookVariableUpdate,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> PlaybookVariableResponse:
    return await service.update_playbook_variable(
        db, variable_id, payload
    )


@router.delete(
    "/variables/{variable_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def delete_playbook_variable(
    variable_id: int,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> None:
    await service.delete_playbook_variable(db, variable_id)


# ------------------------------------------------------------------ #
# Playbook Execution                                                   #
# ------------------------------------------------------------------ #


@router.get(
    "/executions", response_model=PlaybookExecutionListResponse
)
async def list_executions(
    playbook_id: int | None = Query(None),
    status_filter: str | None = Query(None, alias="status"),
    limit: int = Query(50, ge=1, le=500),
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> PlaybookExecutionListResponse:
    return await service.get_executions(
        db, playbook_id, status_filter, limit
    )


@router.get(
    "/executions/{execution_id}",
    response_model=PlaybookExecuteResponse,
)
async def get_execution(
    execution_id: int,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> PlaybookExecuteResponse:
    return await service.get_execution_by_id(db, execution_id)


@router.post(
    "/playbooks/{playbook_id}/execute",
    response_model=PlaybookExecuteResponse,
)
async def execute_playbook(
    playbook_id: int,
    payload: PlaybookExecuteRequest,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> PlaybookExecuteResponse:
    return await service.execute_playbook(db, playbook_id, payload)


@router.post(
    "/playbooks/{playbook_id}/dry-run",
    response_model=DryRunResponse,
)
async def dry_run_playbook(
    playbook_id: int,
    payload: DryRunRequest,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> DryRunResponse:
    return await service.dry_run_playbook(db, playbook_id, payload)


@router.post(
    "/executions/{execution_id}/rollback",
    response_model=RollbackResponse,
)
async def rollback_execution(
    execution_id: int,
    payload: RollbackRequest,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> RollbackResponse:
    return await service.rollback_execution(
        db, execution_id, payload
    )


# ------------------------------------------------------------------ #
# Execution Logs                                                       #
# ------------------------------------------------------------------ #


@router.get(
    "/executions/{execution_id}/logs",
    response_model=ExecutionLogListResponse,
)
async def get_execution_logs(
    execution_id: int,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> ExecutionLogListResponse:
    return await service.get_execution_logs(db, execution_id)


# ------------------------------------------------------------------ #
# Approval Workflows                                                   #
# ------------------------------------------------------------------ #


@router.get(
    "/playbooks/{playbook_id}/workflows",
    response_model=ApprovalWorkflowListResponse,
)
async def list_approval_workflows(
    playbook_id: int,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> ApprovalWorkflowListResponse:
    return await service.get_approval_workflows(db, playbook_id)


@router.post(
    "/playbooks/{playbook_id}/workflows",
    status_code=status.HTTP_201_CREATED,
    response_model=ApprovalWorkflowResponse,
)
async def create_approval_workflow(
    playbook_id: int,
    payload: ApprovalWorkflowCreate,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> ApprovalWorkflowResponse:
    return await service.create_approval_workflow(
        db, playbook_id, payload
    )


@router.put(
    "/workflows/{workflow_id}",
    response_model=ApprovalWorkflowResponse,
)
async def update_approval_workflow(
    workflow_id: int,
    payload: ApprovalWorkflowUpdate,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> ApprovalWorkflowResponse:
    return await service.update_approval_workflow(
        db, workflow_id, payload
    )


@router.delete(
    "/workflows/{workflow_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def delete_approval_workflow(
    workflow_id: int,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> None:
    await service.delete_approval_workflow(db, workflow_id)


# ------------------------------------------------------------------ #
# Approval Requests                                                    #
# ------------------------------------------------------------------ #


@router.get(
    "/approvals/pending",
    response_model=ApprovalRequestListResponse,
)
async def list_pending_approvals(
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> ApprovalRequestListResponse:
    return await service.get_pending_approvals(db)


@router.get(
    "/executions/{execution_id}/approvals",
    response_model=ApprovalRequestListResponse,
)
async def list_execution_approvals(
    execution_id: int,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> ApprovalRequestListResponse:
    return await service.get_approval_requests(db, execution_id)


@router.put(
    "/approvals/{request_id}/approve",
    response_model=ApprovalRequestResponse,
)
async def approve_request(
    request_id: int,
    payload: ApprovalAction,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> ApprovalRequestResponse:
    return await service.approve_request(db, request_id, payload)


@router.put(
    "/approvals/{request_id}/reject",
    response_model=ApprovalRequestResponse,
)
async def reject_request(
    request_id: int,
    payload: ApprovalAction,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> ApprovalRequestResponse:
    return await service.reject_request(db, request_id, payload)


# ------------------------------------------------------------------ #
# Audit Trail                                                          #
# ------------------------------------------------------------------ #


@router.get(
    "/audit", response_model=AuditTrailListResponse
)
async def list_audit_trail(
    entity_type: str | None = Query(None),
    action: str | None = Query(None),
    actor: str | None = Query(None),
    limit: int = Query(100, ge=1, le=1000),
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> AuditTrailListResponse:
    return await service.get_audit_trail(
        db, entity_type, action, actor, limit
    )


# ------------------------------------------------------------------ #
# Schedules                                                            #
# ------------------------------------------------------------------ #


@router.get(
    "/schedules", response_model=PlaybookScheduleListResponse
)
async def list_schedules(
    playbook_id: int | None = Query(None),
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> PlaybookScheduleListResponse:
    return await service.get_schedules(db, playbook_id)


@router.post(
    "/playbooks/{playbook_id}/schedules",
    status_code=status.HTTP_201_CREATED,
    response_model=PlaybookScheduleResponse,
)
async def create_schedule(
    playbook_id: int,
    payload: PlaybookScheduleCreate,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> PlaybookScheduleResponse:
    return await service.create_schedule(db, playbook_id, payload)


@router.put(
    "/schedules/{schedule_id}",
    response_model=PlaybookScheduleResponse,
)
async def update_schedule(
    schedule_id: int,
    payload: PlaybookScheduleUpdate,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> PlaybookScheduleResponse:
    return await service.update_schedule(db, schedule_id, payload)


@router.delete(
    "/schedules/{schedule_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def delete_schedule(
    schedule_id: int,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> None:
    await service.delete_schedule(db, schedule_id)


# ------------------------------------------------------------------ #
# Event Triggers                                                       #
# ------------------------------------------------------------------ #


@router.get(
    "/triggers", response_model=EventTriggerListResponse
)
async def list_triggers(
    playbook_id: int | None = Query(None),
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> EventTriggerListResponse:
    return await service.get_triggers(db, playbook_id)


@router.post(
    "/playbooks/{playbook_id}/triggers",
    status_code=status.HTTP_201_CREATED,
    response_model=EventTriggerResponse,
)
async def create_trigger(
    playbook_id: int,
    payload: EventTriggerCreate,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> EventTriggerResponse:
    return await service.create_trigger(db, playbook_id, payload)


@router.put(
    "/triggers/{trigger_id}",
    response_model=EventTriggerResponse,
)
async def update_trigger(
    trigger_id: int,
    payload: EventTriggerUpdate,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> EventTriggerResponse:
    return await service.update_trigger(db, trigger_id, payload)


@router.delete(
    "/triggers/{trigger_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def delete_trigger(
    trigger_id: int,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> None:
    await service.delete_trigger(db, trigger_id)


@router.post("/triggers/fire/{event_type}")
async def fire_event(
    event_type: str,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> dict:
    results = await service.fire_event(db, event_type)
    return {
        "event_type": event_type,
        "playbooks_triggered": len(results),
    }


# ------------------------------------------------------------------ #
# Clone / Export / Import                                             #
# ------------------------------------------------------------------ #


@router.post("/playbooks/{playbook_id}/clone")
async def clone_playbook(
    playbook_id: int,
    name: str | None = Query(None, description="Name for the cloned playbook"),
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> PlaybookResponse:
    return await service.clone_playbook(db, playbook_id, name)


@router.get("/playbooks/{playbook_id}/export")
async def export_playbook(
    playbook_id: int,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> dict:
    return await service.export_playbook(db, playbook_id)


@router.post("/playbooks/import", status_code=status.HTTP_201_CREATED)
async def import_playbook(
    payload: dict,
    db: Session = Depends(get_db),
    service: AutomationService = Depends(get_automation_service),
) -> PlaybookResponse:
    return await service.import_playbook(db, payload)
